src/backup.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Provides backup functionality for Synapse."""
+import logging
+from typing import Any, Optional
+
+import boto3
+import ops
+from botocore.exceptions import BotoCoreError, ClientError
+from botocore.exceptions import ConnectionError as BotoConnectionError
+from charms.data_platform_libs.v0.s3 import CredentialsChangedEvent, S3Requirer
+from ops.framework import Object
+from pydantic import BaseModel, Field, validator
+
+logger = logging.getLogger(__name__)
+
+S3_CANNOT_ACCESS_BUCKET = "Backup: S3 bucket does not exist or cannot be accessed"
+S3_INVALID_CONFIGURATION = "Backup: S3 configuration is invalid"
+BACK_UP_STATUS_MESSAGES = (S3_INVALID_CONFIGURATION, S3_CANNOT_ACCESS_BUCKET)
+
+
+class S3Parameters(BaseModel):
+    """Configuration for accessing S3 bucket.
+
+    Attributes:
+        access_key: AWS access key.
+        secret_key: AWS secret key.
+        region: The region to connect to the object storage.
+        bucket: The bucket name.
+        endpoint: The endpoint used to connect to the object storage.
+        path: The path inside the bucket to store objects.
+        s3_uri_style: The S3 protocol specific bucket path lookup type.
+    """
+
+    access_key: str = Field(alias="access-key")
+    secret_key: str = Field(alias="secret-key")
+    region: Optional[str] = Field(default=None)
+    bucket: str
+    endpoint: Optional[str]
+    path: str = Field(default="")
+    s3_uri_style: str = Field(alias="s3-uri-style", default="host")
```

Comment:
s/s3_uri_style/s3_uri?

Reply:
Not sure that would be clear, as a S3 URI is something like [S3://bucket-name/key-name](https://repost.aws/questions/QUFXlwQxxJQQyg9PMn2b6nTg/what-is-s3-uri-in-simple-storage-service)

Reply:
So, bucket_pth or smth like that?

Reply:
Makes sense